### PR TITLE
Grt: Congestion report file generation

### PR DESCRIPTION
**Pull request by luis201420**

### CODE REVIEW

**Comment by maliberty** [comment]
@arlpetergadfort FYI based on your request

**Comment by maliberty** [comment]
@luis201420 did you resolve the warnings generated by the GUI?

**Comment by luis201420** [comment]
> @luis201420 did you resolve the warnings generated by the GUI?

Yes, I resolved it

File: src/grt/src/GlobalRouter.cpp [hunk 1]
```diff
@@ -95,6 +95,7 @@ GlobalRouter::GlobalRouter()
       allow_congestion_(false),
       macro_extension_(0),
       verbose_(false),
+      save_congestion_(false),
```
Comment (maliberty) [inline src/grt/src/GlobalRouter.cpp:98]: You should initialize congestion_file_name_(nullptr)

File: src/grt/include/grt/GlobalRouter.h [hunk 2]
```diff
@@ -395,6 +397,10 @@ class GlobalRouter
   float caps_perturbation_percentage_;
   int perturbation_amount_;
 
+  // variables congestion report file
+  bool save_congestion_;
+  const char * congestion_file_name_;
```
Comment (maliberty) [inline src/grt/include/grt/GlobalRouter.h:402]: save_congestion_ is just congestion_file_name_ != nullptr so I don't think you need it.

File: src/grt/src/GlobalRouter.cpp [hunk 3]
```diff
@@ -202,6 +203,34 @@ void GlobalRouter::applyAdjustments(int min_routing_layer,
   fastroute_->initAuxVar();
 }
 
+void GlobalRouter::saveCongestion(){
+  std::vector<std::pair<GSegment, TileCongestion>> congestionGridsV, congestionGridsH;
+  fastroute_->getCongestionGrid(congestionGridsV, congestionGridsH);
+  remove(congestion_file_name_); 
+  std::ofstream out;
+  out.open(congestion_file_name_, std::ios::app);
```
Comment (maliberty) [inline src/grt/src/GlobalRouter.cpp:211]: Just std::ofstream out(congestion_file_name_);

File: src/grt/src/GlobalRouter.cpp [hunk 4]
```diff
@@ -202,6 +203,34 @@ void GlobalRouter::applyAdjustments(int min_routing_layer,
...
+  for (auto & it : congestionGridsH){
```
Comment (maliberty) [inline src/grt/src/GlobalRouter.cpp:212]: clang-format

File: src/grt/src/GlobalRouter.cpp [hunk 5]
```diff
@@ -202,6 +203,34 @@ void GlobalRouter::applyAdjustments(int min_routing_layer,
...
+     out << "violation type: Horizontal congestion\n";
+     int capacity = it.second.first;
+     int usage = it.second.second;
```
Comment (maliberty) [inline src/grt/src/GlobalRouter.cpp:215]: const 
here and below

File: src/grt/src/GlobalRouter.cpp [hunk 6]
```diff
@@ -202,6 +203,34 @@ void GlobalRouter::applyAdjustments(int min_routing_layer,
...
+  }
+  for (auto & it : congestionGridsV){
+     out << "violation type: Vertical congestion\n";
+     int capacity = it.second.first;
+     int usage = it.second.second;
+     out << "\tsrcs: capacity:" << capacity << " usage:" << usage << " overflow:" <<usage-capacity << "\n";
+     odb::Rect rect = globalRoutingToBox(it.first);
+     out << "\tbbox = ";
+     out << "( " << dbuToMicrons(rect.xMin()) << ", " << dbuToMicrons(rect.yMin()) << " ) - ";
+     out << "( " << dbuToMicrons(rect.xMax()) << ", " << dbuToMicrons(rect.yMax()) << ") on Layer \n";
```
Comment (maliberty) [inline src/grt/src/GlobalRouter.cpp:230]: why are you printing "on Layer" ?

File: src/gui/src/drcWidget.cpp [hunk 7]
```diff
@@ -466,6 +467,10 @@ void DRCWidget::loadTRReport(const QString& filename)
                      line);
     }
 
+    if (type == "Horizontal congestion" || type == "Vertical congestion") {
+      is_congestion_file = true;
+    }
```
Comment (maliberty) [inline src/gui/src/drcWidget.cpp:472]: Rather than make this congestion specific just allow a layer value like "-" for no value

**Review by maliberty — CHANGES_REQUESTED** [review]

**Review by eder-matheus — APPROVED** [review]
@maliberty I've helped Luis with clang-format, and I think your other comments were also addressed. I didn't find anything else, so I'm approving it. But we can wait until your final review.

**Review by maliberty — APPROVED** [review]

**Pull request merged**